File: sdk_core/src/main/java/com/pandaq/appcore/browser/ProcessWebView.java
package com.pandaq.appcore.browser;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Log;
import android.view.ViewGroup;
import android.webkit.*;
import android.widget.ProgressBar;

import com.pandaq.appcore.BuildConfig;
import com.pandaq.appcore.R;


/**
 * Created by PandaQ on 2017/6/29.
 * 带进度条的WebView
 */

public class ProcessWebView extends WebView {

    private ProgressBar mProgressBar;

    public ProcessWebView(Context context) {
        this(context, null);
    }

    public ProcessWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mProgressBar = new ProgressBar(context, null, android.R.attr.progressBarStyleHorizontal);
        mProgressBar.setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, 10, 0, 0));
        Drawable drawable = context.getResources().getDrawable(R.drawable.core_webview_process_state, null);
        mProgressBar.setProgressDrawable(drawable);
        addView(mProgressBar);
        setWebViewClient(new MyWebClient());
        setWebChromeClient(new ChromeClient());
        setWebContentsDebuggingEnabled(BuildConfig.DEBUG);
    }

    private class ChromeClient extends WebChromeClient {
        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            super.onProgressChanged(view, newProgress);
            if (newProgress == 100) {
                mProgressBar.setVisibility(GONE);
            } else {
                if (mProgressBar.getVisibility() == GONE) {
                    mProgressBar.setVisibility(VISIBLE);
                    mProgressBar.setProgress(newProgress);
                }
            }
        }

    }

    private class MyWebClient extends WebViewClient {

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            view.loadUrl(request.getUrl().toString());
            return true;
        }

        @Override
        public void onPageFinished(WebView view, String url) {
            super.onPageFinished(view, url);
            post(() -> {
                view.measure(0, 0);
                getLayoutParams().height = view.getMeasuredHeight();
            });

        }
    }

}

File: sdk_core/src/main/java/com/pandaq/appcore/browser/ProcessWebView.kt
package com.pandaq.appcore.browser

import android.content.Context
import android.graphics.Bitmap
import android.os.Build
import android.util.AttributeSet
import com.pandaq.appcore.BuildConfig
import com.pandaq.appcore.browser.bridge.BridgeData
import com.pandaq.appcore.browser.bridge.JavaScriptApis
import com.pandaq.rxpanda.utils.GsonUtil
import com.tencent.smtt.export.external.interfaces.WebResourceRequest
import com.tencent.smtt.sdk.WebChromeClient
import com.tencent.smtt.sdk.WebView
import com.tencent.smtt.sdk.WebViewClient

/**
 * Created by PandaQ on 2017/6/29.
 * 带进度条的WebView
 */
class ProcessWebView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) :
    WebView(context, attrs) {

    private var loadCallback: LoadCallback? = null

    init {
        webViewClient = MyWebClient()
        webChromeClient = ChromeClient()
        addJavascriptInterface(JavaScriptApis(), "Android")
        setWebContentsDebuggingEnabled(BuildConfig.IN_DEBUG)
    }

    private inner class ChromeClient : WebChromeClient() {
        override fun onProgressChanged(view: WebView, newProgress: Int) {
            super.onProgressChanged(view, newProgress)
        }
    }

    private inner class MyWebClient : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            return if (Build.VERSION.SDK_INT < 26) {
                view.loadUrl(request.url.toString())
                true
            } else {
                false
            }
        }

        override fun onLoadResource(p0: WebView?, p1: String?) {
            super.onLoadResource(p0, p1)
            loadCallback?.onLoadResource()
        }

        override fun onPageCommitVisible(p0: WebView?, p1: String?) {
            super.onPageCommitVisible(p0, p1)
        }

        override fun onPageStarted(view: WebView?, url: String?, p2: Bitmap?) {
            super.onPageStarted(view, url, p2)
            loadCallback?.onStart()
        }

        override fun onPageFinished(view: WebView, url: String) {
            super.onPageFinished(view, url)
            loadCallback?.onFinish()
        }
    }

    /**
     * 向 H5 端发送数据
     *
     *
     */
    fun sendDataToH5(data: BridgeData) {
        val script = "javascript:${data.methodName}(${GsonUtil.gson().toJson(data.data)})"
        loadUrl(script)
    }

    override fun canGoBack(): Boolean {
        return super.canGoBack() && !url.contains("homepage")
    }

    fun setLoadCallback(loadCallback: LoadCallback?) {
        this.loadCallback = loadCallback
    }

    interface LoadCallback {
        fun onStart()
        fun onLoadResource()
        fun onFinish()
    }
}